Rework mutable namespace resolution to handle recursion

wking opened the pull request:
There are detailed notes in the individual commits, but the meat is in
the opening "namesys: Add recursive resolution".  This is a work in
progress towards the separation sketched out [here][1] and endorsed
[here][2].  For example, with this PR you can:

    $ ipfs dns ipfs.tremily.us
    /dnslink/tremily.us
    $ ipfs dns tremily.us
    /dnslink/tremily.us
    $ ipfs dns --recursive ipfs.tremily.us
    /ipns/QmbqDJaoZYoGNw4dz7FqFDCf6q9EcHoMBQtoGViVFw2qv7

However, I'm currently sticking on the multihash encoding.  The
/ipns/… and /ipfs/… multihashes are base-58, but there's also the
proquint encoding.  However, proquint isn't a *protocol*, you can have
proquint-encoded multihashes in the IPNS and IPFS protocol spaces.
How do we distinguish that?  The heavy approach is:

  /ipfs-proquint/…
  /ipns-proquint/…

in which case maybe we want to be using:

  /ipns-base-58/…
  /ipfs-base-58/…

where we currently use /ipns/… and /ipfs/….  Thoughts?  More clever
solutions?  Should the encoding-choice be part of a layer between
binary multihashes and the IP*S libraries?  For example, maybe encoded
multihashes should have a leading character or two representing their
encoding (hex “a:1220…”, base-58 “b:Qm…”, proquint
“c:bikad-fokor-zihoj-bajir”, …).

Also, do we want to use /dnslink/… (to match dnslink=…) or /dns/…
(because “link” is mostly wasted space once we're in the context of
resolvable names)?

[1]: https://github.com/ipfs/go-ipfs/issues/1082#issuecomment-99255884
[2]: https://github.com/ipfs/go-ipfs/issues/1082#issuecomment-99417421

wking commented:
This is a farily major resolution overhaul, so we probably need to figure out how it relates to the #1150 fix @cryptix [is working on][1].

[1]: https://github.com/ipfs/go-ipfs/issues/1150#issuecomment-97070601

cryptix commented:
Hey @wking,

thanks for going through this! I also thought similar things when working on #1150. I actually wouldn't mind for this going in first but don't need it.

One thing i'm not too sure about is exposing the `depth int` recursion parameter. Would would it mean to recurse with `42`? If it's function is to specify to wich protocol level the use wants to resolve i'd like something like this much more:


```go
type Protocol uint
const (
	DNS Protocol = iota
	IPNS
	IPFS
)

type Resolver interface {
	Resolve(ctx context.Context, name string, proto Protocol) (path.Path, error)
}
```

wking commented:
On Thu, May 07, 2015 at 04:35:11PM -0700, Henry wrote:
> One thing i'm not too sure about is exposing the `depth int`
> recursion parameter. Would would it mean to recurse with `42`?

“Please resolve this recursively, but bail out after 32 cycles”.  This
lets you do things like the current maxLinks stuff [1,2], although I'd
rather be setting that limit via a config variable (and as this PR
stands, the limit is hardcoded to the infinite 0 in several places).

> If it's function is to specify to wich protocol level the use wants
> to resolve…

Nope.  The per-protocol handlers currently just take the bare object.
For example:

  DNSResolver.Resolve(ctx, "ipfs.io", 0)

and the multi-protocol resolver gets the protocol from a prefix:

  NewNameSystem(…).Resolve(ctx, "/dnslink/ipfs.io", 0)

[1]: https://github.com/ipfs/go-ipfs/blob/bf5b8032a802b98589d1e3bef1508a9d80c77699/core/pathresolver.go#L12
[2]: https://github.com/ipfs/go-ipfs/blob/bf5b8032a802b98589d1e3bef1508a9d80c77699/core/pathresolver.go#L24

wking commented:
I've rebased this onto #1209 to get my local test suite going again,
and squashed in a few more fixes, changed ‘ipfs name resolve …’ to be
IPNS-only, and shifted the multi-protocol version to ‘ipfs resolve …’.

Besides the encoding issues (base-58 vs. hex vs. proquint) mentioned
in my lead-off comment, I'm also concerned about FUSE mount handling.
Currently /ipns/… is back to IPNS-only handling (e.g. you can't use
/ipns/example.com/).  We *could* pass the first two segments of the
post-mount-point path through to use the new explicit-prefix resolver
(e.g. /ipns/dnslink/example.com/), but that makes the /ipns/ bit seem
strange.  Another alternative is to have a separate root directory for
each protocol (/ipfs/…, /ipns/…, /dnslink/…, …), but that seems ugly
too.  How about /ipr/ipfs/…, /ipr/ipns/…, /ipr/dnslink/…, where ipr
stands for “InterPlanetary Resolver”?

jbenet commented:
@wking this is looking really good! I'll sink into CR in a bit. Do you want it to be merged after or before @cryptix's stuff? Whatever is easiest for you both.

jbenet commented:
Let's use `/dns/...`.

---

On the encoding, speaking generally for a moment: [it's become apparent to me](https://github.com/jbenet/random-ideas/issues/17) that we need some sort of self-describing encoding. It definitely _is_ a problem that `/ipfs` and `/ipns` links use base58 only. (as you point out, what does it mean to have `/ipfs/<hex>`?). I think this is --again-- a place where we want to use self-description to signal the encoding. 

In the IPNS case specifically-- for chosing between proquint and other things, the rules we're using now do expression matching (i.e. if there's a `(<word>-)+<word>` pattern it's a proquint, if there's a domain it's dns (+ .bit/.onion "tlds"). else it's a hash. This _works_ for the most part.

I do not want to use a double prefix like `/ipr` because (a) `/ipfs` and `/ipns` is already a lot-- these are not just paths, they're URLs. There's a constraint here where the web addresses and the unix pathnames should be the same, so that users can freely enter the exact same string in both (something we cannot do today). and (b), i want to push things towards another idea, [fs-root](https://github.com/jbenet/random-ideas/issues/28) (moving unix filesystems towards mounting lots of protocols). 

So the solution i'd prefer in the abstract is to have `/dns/...` be a top level. I think this definitely would make sense for a `dnslink` tool. but think it is clunky in the ipfs toolchain because now we're asking for _3_ directories at root. there is also the fact that we'd like `/ipns/` links to work for users, with dns being a bit of a hack. (so `/ipns/` is acting like a resolver of dns names, and ipns pki records). I think in the dns cases, the expression matching works well in practice-- domains and hashes are distinctive enough. proquint is sort of an encoding, so maybe we should be treating it like hex (maybe a self-describing prefix?). -- the gist is that i agree it's impure, but not sure we can expect the world to go full on to "fs root with all the protocol prefixes \o/" land just yet.

jbenet commented:
I merged #1189 first, so let's rebase this one on top.

jbenet commented on core/commands/dns.go:
```diff
@@ -0,0 +1,82 @@
+	util "github.com/ipfs/go-ipfs/util"
+)
+
+var DNSCmd = &cmds.Command{
```
+1: to this command :)

jbenet commented on core/commands/ipns.go:
```diff
@@ -0,0 +1,104 @@
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+var ipnsCmd = &cmds.Command{
```
very minor thing-- i think we'll later want this to be something like `ipnsResolve` (later, doesnt matter), because i think `ipfs name` and `ipns` can be the same thing. (i.e. `ipnsCmd` would have both `publish` and `resolve` as children). thus an `ipns` independent tool could publish a record.

jbenet commented on core/commands/resolve.go:
```diff
@@ -14,37 +13,46 @@ type ResolvedPath struct {
+		cmds.StringArg("name", true, false, "The name to resolve.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("recursive", "r", "Resolve until the result is an IPFS name"),
```
maybe this should default to true?

jbenet replied:
no, i think you're right. it should default to false.

jbenet commented on core/commands/resolve.go:
```diff
@@ -62,27 +70,19 @@ Resolve te value of another name:
-				res.SetError(errors.New("Identity not loaded!"), cmds.ErrNormal)
-				return
-			}
-			name = n.Identity.Pretty()
```
no longer default to the local identity? this is probably ok for `resolve`, but may feel asymmetric for users of `publish`

jbenet replied:
ahhh nevermind, this is just `ipfs resolve` now. not `ipfs name resolve`

jbenet commented on core/commands/resolve.go:
```diff
@@ -62,27 +70,19 @@ Resolve te value of another name:
 			return
 		}
 
-		// TODO: better errors (in the case of not finding the name, we get "failed to find any peer in table")
```
do we return better errors now?

wking replied:
On Thu, May 07, 2015 at 10:51:49PM -0700, Juan Batiz-Benet wrote:
> > -		// TODO: better errors (in the case of not finding the name, we get "failed to find any peer in table")
> 
> do we return better errors now?

Without a daemon:

  $ ipfs /ipfs/QmNonExistentKey
  panic?
  $ ipfs resolve /ipns/QmNonExistentKey
  Error: datastore: key not found
  $ ipfs resolve /dns/abc
  Error: not a valid domain name
  $ ipfs resolve /dns/example.com
  Error: could not resolve name.

With a daemon running, the DNS errors are the same, and the IPFS/IPNS
errors are:

  $ ipfs /ipfs/QmNonExistentKey
  Error: context deadline exceeded
  $ ipfs resolve /ipns/QmNonExistentKey
  Error: routing: not found

What sort of better errors were you looking for?

jbenet replied:
Hmm-- we probably should make them the same.

This is odd:
```
> ipfs /ipfs/QmNonExistentKey
Error: context deadline exceeded
```
cmdslib should just error out...

jbenet replied:
```
> ipfs resolve /ipns/QmNonExistentKey
Error: routing: not found
```

should probably be

```
> ipfs resolve /ipns/QmNonExistentKey
Error: could not resolve name.
```
in both cases.

wking commented:
On Thu, May 07, 2015 at 10:25:33PM -0700, Juan Batiz-Benet wrote:
> I do not want to use a double prefix like `/ipr` because (a) `/ipfs`
> and `/ipns` is already a lot-- these are not just paths, they're
> URLs.

Well, the URL forms will have some other prefix in front
(e.g. http://gateway.ipfs.io/).  I think a local filesystem prefix
like /ipr/ is similar (you can cut and paste between the two once you
know both prefixes), and it allows us to make as many per-protocol
directories as we want without needing lots of root-permission steps.
But this IPR-mount-point-prefix should be configurable anyway (and
might already be), so maybe *I* should just tweak my local config to
use /ipr/… ;).  Then we/I can remove the /ipr/ prefix once the
root-prefixed protocol paths get better traction.

I'll look over the rest of your references and rebase onto the current
master tomorrow.

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -81,7 +81,7 @@ func (i *gatewayHandler) resolveNamePath(ctx context.Context, p string) (string,
 		elements := strings.Split(p[len(IpnsPathPrefix):], "/")
 		hash := elements[0]
-		rp, err := i.node.Namesys.Resolve(ctx, hash)
+		rp, err := i.node.Namesys.Resolve(ctx, hash, 0)
```
don't think we should resolve names infinitely. we should have a max, something like `16` or `32`. otherwise we can hose machines by setting up very long chains and causing nodes to resolve for a very long time). (maybe infinitely, if we can generate + publish names faster than the nodes time out)

wking replied:
On Thu, May 07, 2015 at 10:54:51PM -0700, Juan Batiz-Benet wrote:
> > @@ -81,7 +81,7 @@ func (i *gatewayHandler) resolveNamePath(ctx context.Context, p string) (string,
> >  	if strings.HasPrefix(p, IpnsPathPrefix) {
> >  		elements := strings.Split(p[len(IpnsPathPrefix):], "/")
> >  		hash := elements[0]
> > -		rp, err := i.node.Namesys.Resolve(ctx, hash)
> > +		rp, err := i.node.Namesys.Resolve(ctx, hash, 0)
> 
> … otherwise we can hose machines by setting up very long chains and
> causing nodes to resolve for a very long time). (maybe infinitely,
> if we can generate + publish names faster than the nodes time out)

Or even just by creating a cycle, which is easy enough with DNS links
now, and will become easy with other protocols once we get support for
IPNS → (IPNS | DNS) and IPFS → (IPNS | DNS).  I just don't like
hard-coding a magic recursion limit.  I'd suggest the usual “configure
per-node in ~/.ipfs/config and override from $IPFS_RECURSION_LIMIT”,
but I haven't had time to look into the config handling.  I'm fine
seeding the config value on ‘ipfs init’ with whatever you think is
reasonable.

jbenet replied:
We'll have to provide recommended bounds in the specs. this is a protocol problem, not an implementation one. The resolution limit picked will impact all sorts of implementations out there. If some implementation works with limit `k`, but others with limit `k/2`, it will create confusion for users. (dns referral / name recursion had a similar problem)

whyrusleeping replied:
> this is a protocol problem, not an implementation one. 

:+1:

jbenet commented on namesys/namesys.go:
```diff
@@ -1,59 +1,95 @@
 )
 
-// ipnsNameSystem implements IPNS naming.
+// ipfs NameSystem implements generic IPFS naming.
```
ah yes. this is a bit of a tricky naming thing because-- yeah, ipns means mostly pki based ptrs, and "ipfs naming" could be defined as having many name resolvers (ipns + dns + ...). So far, it's been most useful to describe it to people as clearly isolating out `/ipfs/...` to be immutable _content-addressed_ stuff, and `/ipns/...` to be mutable name-resolved stuff. (agree that it's a bit unclean, but am not sure the naming switch helps. an alternative semantic construction may be to say that "ipns supports sfs-style pki based names, and other resolvers (dns + ...) for doing things like pretty names and so on."

wking replied:
On Thu, May 07, 2015 at 11:03:11PM -0700, Juan Batiz-Benet wrote:
> > -// ipnsNameSystem implements IPNS naming.
> > +// ipfs NameSystem implements generic IPFS naming.
> 
> ah yes. this is a bit of a tricky naming thing because-- yeah, ipns
> means mostly pki based ptrs, and "ipfs naming" could be defined as
> having many name resolvers (ipns + dns + ...). So far, it's been
> most useful to describe it to people as clearly isolating out
> `/ipfs/...` to be immutable _content-addressed_ stuff, and
> `/ipns/...` to be mutable name-resolved stuff. (agree that it's a
> bit unclean, but am not sure the naming switch helps. an alternative
> semantic construction may be to say that "ipns supports sfs-style
> pki based names, and other resolvers (dns + ...) for doing things
> like pretty names and so on."

That's a bit long for a godoc description ;).  The “ipfs” here is
“understood by this resolver from the go-ipfs repository”.  Maybe just
coin a new catch-all for mutable names: “mutable NameSystem implements
all IPFS naming protocols…”?

jbenet replied:
this is not about godoc or the ipfs implementations-- it's about how we describe the system that resolves:

```
/ipns/Qm..../...
/ipns/foo.com/...
```
It comes to bear on the end users, who should be required to learn the least possible amount. Yeah i corroborate my earlier comments-- this should stay defined as IPNS. it's less pure than inventing another abstraction, but the description:

> "ipns supports sfs-style pki based names, and other resolvers (dns + ...)."

is pure enough and gives end users less to learn. Until a day in which `fsroot` is trivial to deploy on the world, let's keep it as it's been: IPNS is the _name system_. it resolves SFS-style PKI names, domains, and more.

wking replied:
On Sat, May 09, 2015 at 09:57:01AM -0700, Juan Batiz-Benet wrote:
> > -// ipnsNameSystem implements IPNS naming.
> > +// ipfs NameSystem implements generic IPFS naming.
> …
> this is not about godoc or the ipfs implementations-- it's about how
> we describe the system that resolves:
> 
> ```
> /ipns/Qm..../...
> /ipns/foo.com/...

I think that's also about the godoc.  We should have a canonical
explanation for what we're about here, and the godoc is a good place
to store that.  If we want a more elaborate form in the docs (maybe in
the tour?) that's fine, but I think we need to clarify what goes into
the godocs before biting that off.

And as I've said earlier [1], I want a separate directory for
/dns/foo.com/… to clearly distinguish between DNS-based links and
DHT-based links.  I think using “IPNS” for both DNS-links and
DHT-links is going to cause more confusion than having a single
directory saves you.  So far, I've been suggesting “IPNS” just apply
to the DHT links, but I'd also be ok with:

  /ipns/dht/Qm…
  /ipns/dns/foo.com…

So what I want:

1. Explicit prefixes/namespaces for the different protocols.  I expect
   this will make the system *less* confusing, by avoiding ambiguity
   about which protocol is resolving which name.  That helps:
   a. Users, by making it easy to identify (and find docs, file bugs,
      …) for the protocol they're using.
   b. Protocol authors, who only need to register a unique
      prefix/namespace for their protocol, and not worry about making
      sure their hex-encoded frobnitz references aren't accidentally
      treated as base-58 multihashes.
   c. Sysadmins, who don't need to worry about resolution order when
      they're configuring the resolver to decide which protocol gets
      to interpret an ambiguous name.

2. An explicit name for SFS-style references stored in the DHT.  I've
   been calling this system “IPNS”, but if that's supposed to be
   mutable links in general, we'll need a different name here (and
   “SFS-style PKI names” isn't an easy-to-search, unambiguous name ;).

3. A clear procedure for protocol authors to integrate their mutable
   name system with the IPFS-world tooling.  This would be trivial
   with FUSE, just return symlinks:

     /frobnitz/frobby-key → /dns/foo.com
     /frobnitz/nitzy-key → /ipfs/QmSomeKey/a/b/c

   For getting into ‘ipfs resolve …’ we'll probably want something
   that the resolver can launch as a subprocess or call as a service
   if it doesn't have a built-in resolver for a given protocol name.

> … it's less pure than inventing another abstraction, but the
> description:
> 
> > "ipns supports sfs-style pki based names, and other resolvers (dns
> > + ...)."
> 
> is pure enough and gives end users less to learn.

I don't think many users will care one way or the other if the mutable
name system is:

  /ipns/<all-the-keys>

or:

  /ipns/<protocol>/<protocol-specific-keys>

or:

  /<protocol>/<protocol-specific-keys>

Personally, I favor the last form to mirror
/ipfs/<ipfs-specific-keys>.  Do you really think it's a big usability
win to drop the protocol namespace?  It makes the implementation so
much trickier…

wking replied:
On Sat, May 09, 2015 at 12:12:13PM -0700, W. Trevor King wrote:
>   /<protocol>/<protocol-specific-keys>
> 
> Personally, I favor the last form to mirror
> /ipfs/<ipfs-specific-keys>.

And it's good to note that what you're storing in the alternative
protocol doesn't have to be a link.  I could store *content* under
/dns/… or in the DHT (/dht/…?).  That's not the greatest idea for
those protocols, but it's how we handle public keys now (#1045), and I
see no reason to rule it out at a protocol/implementation level.  This
also makes it easy to interact with non-IPFS immutable protocols, e.g.:

  /some-other-content-addressable-system/key-into-their-space/…

I don't see a need for the protocol cross-links to distinguish between
“this is a link into the immutable IPFS space” or “this is a link into
the mutable DNS space that the IPNS resolver understands”.  All it
needs to know is “the key I was given resolves to a reference to key
$X in protocol $Y”.

jbenet commented on namesys/namesys.go:
```diff
@@ -1,59 +1,95 @@
+	return &ipfs{
+		resolvers: map[string]Resolver{
+			"/dnslink/": new(DNSResolver),
+			"/proquint/": new(ProquintResolver),
```
:-1: let's not make proquints a protocol-- as we both mentioned above, proquint is more of an encoding. so maybe we should be prefixing it (if the text doesnt follow an _always predictable_ pattern)

jbenet commented:
> Well, the URL forms will have some other prefix in front (e.g. http://gateway.ipfs.io/).

this is a stop-gap:
- it will shortly be just `http://ipfs.io/`
- I hope to get a tld (so `/ipfs/...` just works)
- the firefox + chrome extensions already resolve paths directly (i believe? or will soon?). and 
- we can convince browsers vendors to implement support for it. 
I'd rather not add another stop-gap (`the /ipr` prefix) where we don't absolutely have to-- i understand the flexibility niceness, but the problem is people are going to start hard-coding `/ipr` into things and that's going to hurt a lot down the road. (this is also why i tell people to link to ipfs with `/ipfs/...` and not with `gateway.ipfs.io/ipfs/...` unless they absolutely have to (they dont control the rendering). In fact, i'll likely start making links like this from now on: [/ipfs/QmSYCpuKPbPQ2iFr2swJj2hvz7wQUXfPBXPiuVsQdL5FEs/cat.jpg](http://gateway.ipfs.io/ipfs/QmSYCpuKPbPQ2iFr2swJj2hvz7wQUXfPBXPiuVsQdL5FEs/cat.jpg)

whyrusleeping commented on core/commands/ipns.go:
```diff
@@ -0,0 +1,104 @@
+			return
+		}
+
+		// TODO: better errors (in the case of not finding the name, we get "failed to find any peer in table")
```
failed to find any peers in table means that bootstrapping was not successful, not that a lookup necessarily resulted in 'not found'

whyrusleeping commented:
@wking status update here?

wking commented:
On Sun, May 10, 2015 at 04:08:26PM -0700, Jeromy Johnson wrote:
> @wking status update here?

After some IRC discussion with @jbenet, I need to reroll this to:

1. Leave the current “try all the protocols” handling in place.  We're
   putting off the migration towards explicit per-protocol prefixes
   for now.
2. Maybe not update the IPNS resolver to be DHT-entry specific (it's
   currently multi-protocol)?

I haven't had time to reroll it yet.  Maybe tonight?  Maybe tomorrow?
I also haven't gotten through a full test run yet, mostly because
something will hang at some point.  I haven't looked into this
carefully enough to know if it's always the same test hanging, or if
it's a collection of random unreliable tests.

wking commented:
On Sun, May 10, 2015 at 08:31:16PM -0700, W. Trevor King wrote:
> After some IRC discussion with @jbenet, I need to reroll this to:
> 
> 1. Leave the current “try all the protocols” handling in place.  We're
>    putting off the migration towards explicit per-protocol prefixes
>    for now.

Done.

> 2. Maybe not update the IPNS resolver to be DHT-entry specific (it's
>    currently multi-protocol)?

Not done yet, and the test suite passes locally, so maybe we don't
need to?  This branch adds ‘ipfs resolve …’ which handles
multi-protocol resolution, and that should be a simple migration for
users.

I also renamed the multi-protocol type to ‘mpns’ for Multi-Protocol
Name System to avoid previous confusion between IPNS for the whole
mutable space or just the DHT links [1].

We still need:

* Docs.  Where should these live?  Is the command-line help sufficent?
  Do I need migration notes about the ‘ipfs name resolve …’ scope
  changes?  Output format changes?
* Possibly more consistent error handling [2].
* A cap on recursive resolution [3].  Maybe hard code a default for
  now (depth == 0 → 32?) and add configurable defaults at the
  ipfs-config and command-line-option levels in follow-up PRs?

[1]: https://github.com/ipfs/go-ipfs/pull/1208/files#r29994332
[2]: https://github.com/ipfs/go-ipfs/pull/1208#discussion_r29918312
[3]: https://github.com/ipfs/go-ipfs/pull/1208#discussion_r29918387

whyrusleeping commented:
the tests probably pass because we only have tests for ipns records... 50% of that is because i have no idea how to test dns resolution in a sandbox

wking commented:
On Tue, May 12, 2015 at 11:40:17PM -0700, Jeromy Johnson wrote:
> 50% of that is because i have no idea how to test dns resolution in
> a sandbox

I'll make net.LookupTXT pluggable, add a mock version, and use the
mock in new namesys/dns_test.go and namesys/namesys_test.go tests.
Will that address your concerns?

whyrusleeping commented:
That sounds good to me

jbenet commented on core/commands/dns.go:
```diff
@@ -0,0 +1,82 @@
+
+The resolver will give:
+
+  > ipfs dns ipfs.io
```
i think this needs to be `ipfs.ipfs.io`

jbenet replied:
may want to use `foo.ipfs.io`

wking replied:
On Thu, May 14, 2015 at 09:21:34PM -0700, Juan Batiz-Benet wrote:
> > +  > ipfs dns ipfs.io
> 
> may want to use `foo.ipfs.io`

So two questions:

1. Does it really matter if this example matches up with live DNS
   records?
2. What live DNS links do we have?  I don't see TXT records for
   ipfs.io, ipfs.ipfs.io, or foo.ipfs.io.

jbenet commented on core/commands/publish.go:
```diff
@@ -102,7 +103,7 @@ Publish an <ipfs-path> to another public key (not implemented):
 		cmds.Text: func(res cmds.Response) (io.Reader, error) {
 			v := res.Output().(*IpnsEntry)
-			s := fmt.Sprintf("Published name %s to %s\n", v.Name, v.Value)
+			s := fmt.Sprintf("Published %s to %s\n", v.Value, v.Name)
```
this order is flipped from the input, if you add the optional entry (which today doesn't work), you'd get:

```
> ipfs name publish <name> <value>
Published <value> to <name>
```

this can be fine, but a heads up that it may be odd in the future.

wking replied:
On Thu, May 14, 2015 at 09:24:16PM -0700, Juan Batiz-Benet wrote:
> > -			s := fmt.Sprintf("Published name %s to %s\n", v.Name, v.Value)
> > +			s := fmt.Sprintf("Published %s to %s\n", v.Value, v.Name)
> 
> this order is flipped from the input…

We're currently not doc'ing this consistently (see the commit message
for 7e9db663, core/commands/publish: Fix published message,
2015-05-07).  To preserve the argv order, how about:

  $ ipfs name publish <name> <value>
  Published to <name>: <value>

?

jbenet replied:
sgtm :+1:

jbenet commented on core/commands/resolve.go:
```diff
@@ -14,37 +13,46 @@ type ResolvedPath struct {
+var ResolveCmd = &cmds.Command{
 	Helptext: cmds.HelpText{
-		Tagline: "Gets the value currently published at an IPNS name",
+		Tagline: "Resolve all supported protocols to IPFS",
```
perhaps `Resolve the value of names`

wking replied:
On Thu, May 14, 2015 at 09:25:54PM -0700, Juan Batiz-Benet wrote:
> >  	Helptext: cmds.HelpText{
> > -		Tagline: "Gets the value currently published at an IPNS name",
> > +		Tagline: "Resolve all supported protocols to IPFS",
> 
> perhaps `Resolve the value of names`

Sure.

jbenet commented on core/commands/resolve.go:
```diff
@@ -14,37 +13,46 @@ type ResolvedPath struct {
-IPNS is a PKI namespace, where names are the hashes of public keys, and
-the private key enables publishing new (signed) values. In resolve, the
-default value of <name> is your own identity public key.
+There are a number of mutable protocols that can link among themselves
```
`... number of mutable name protocols ...`

wking replied:
On Thu, May 14, 2015 at 09:26:10PM -0700, Juan Batiz-Benet wrote:
> >  		ShortDescription: `
> > -IPNS is a PKI namespace, where names are the hashes of public keys, and
> > -the private key enables publishing new (signed) values. In resolve, the
> > -default value of <name> is your own identity public key.
> > +There are a number of mutable protocols that can link among themselves
> 
> `... number of mutable name protocols ...`

Added to my current reroll.

jbenet commented on core/pathresolver.go:
```diff
@@ -11,64 +10,42 @@ import (
 
 		segments := append(respath.Segments(), extensions...)
-		r.p, err = path.FromSegments(segments...)
+		p, err = path.FromSegments("/", segments...)
```
this process (`split path, resolve first 2 components, re-base path`) happens often, may want a function for it in `path` pkg

wking replied:
On Thu, May 14, 2015 at 09:30:36PM -0700, Juan Batiz-Benet wrote:
> > -		r.p, err = path.FromSegments(segments...)
> > +		p, err = path.FromSegments("/", segments...)
> 
> this process (`split path, resolve first 2 components, re-base
> path`) happens often, may want a function for it in `path` pkg

I think a better approach is something like:

1. Split path into segments to get a path stack.
2. Inject a leading ‘ipfs’ segment if the initial path didn't start
   with a slash.
3. Initialize (or clear) the cwd array.
4. First segment selects the protocol (currently ‘ipfs’ or ‘ipns’),
   second segment selects the name we need to resolve.
5. Pop the second segment from the path stack and resolve it using the
   protocol from the first segment.  The result is:
   a. A symlink to an absolute path.
      i. Split the referenced path into segments.
      ii. Inject a leading ‘ipfs’ segment if the referenced path
          didn't start with a slash.
      iii. Append the trailing components from the initial path to the
           referenced path.
      iv. Go to (3).
   b. A symlink to a relative path.
      i. Split the referenced path into segments.
      ii. Pop leading ‘..’ from the referenced segments.  For each
          ‘..’ popped, pop an entry off the cwd array, erroring out if
          the cwd array is empty.
      iii. Prefix the protocol segment to the remaining referenced
           segments, and then append the trailing components from the
           path stack to the remaining referenced segments.
             ‘[protocol] + [symlink, stuff] + [stack, stuff]’
      iv. Go to (4).
   c. An object with children.
      i. Add the object to the cwd stack.
      ii. Go to (4).

The fact that we currently only manipulate the first two components
during non-IPFS resolution is part of what makes it hard to link back
to IPNS-space (DHT links or DNS links) from IPFS-space.

But getting generic symlink-or-not handling in the resolver doesn't
have to land in this PR ;).

jbenet commented on namesys/routing.go:
```diff
@@ -30,15 +30,29 @@ func NewRoutingResolver(route routing.IpfsRouting) Resolver {
-// names.
-func (r *routingResolver) Resolve(ctx context.Context, name string) (path.Path, error) {
+// Resolve implements Resolver.
+func (r *routingResolver) Resolve(ctx context.Context, name string, depth int) (path.Path, error) {
```
i dont think we should encourage users to chose different resolution depths. it may be useful in some odd cases, but i want to minimize the variability of this as it is an important protocol concern.

whyrusleeping replied:
:+1:

wking replied:
On Thu, May 14, 2015 at 09:32:03PM -0700, Juan Batiz-Benet wrote:
> > -func (r *routingResolver) Resolve(ctx context.Context, name string) (path.Path, error) {
> > +// Resolve implements Resolver.
> > +func (r *routingResolver) Resolve(ctx context.Context, name string, depth int) (path.Path, error) {
> 
> i dont think we should encourage users to chose different resolution
> depths. it may be useful in some odd cases, but i want to minimize
> the variability of this as it is an important protocol concern.

So the only cases you want to expose are “resolve a single step” and
“resolve as many steps as you can without violating some configured
limit”?  I'm ok with that, but I'd like a clearer picture of how that
configured limit gets set and managed before taking this knob off.

jbenet replied:
yeah i think so. the limit could be a constant on the Resolver when it is initialized.

jbenet replied:
(we _could_ provide this function (with depth) as a separate one, but mostly encourage use of the one without the param)

jbenet commented on fuse/ipns/ipns_unix.go:
```diff
@@ -140,7 +140,7 @@ func (s *Root) Lookup(ctx context.Context, name string) (fs.Node, error) {
 
 	// other links go through ipns resolution and are symlinked into the ipfs mountpoint
-	resolved, err := s.Ipfs.Namesys.Resolve(s.Ipfs.Context(), name)
+	resolved, err := s.Ipfs.Namesys.Resolve(s.Ipfs.Context(), name, 0)
```
why is this resolving with depth 0? nothing should resolve infinitely. yeah this is what i want to avoid

jbenet replied:
(or am i forgetting something?)

wking replied:
On Thu, May 14, 2015 at 09:33:18PM -0700, Juan Batiz-Benet wrote:
> > -	resolved, err := s.Ipfs.Namesys.Resolve(s.Ipfs.Context(), name)
> > +	resolved, err := s.Ipfs.Namesys.Resolve(s.Ipfs.Context(), name, 0)
> 
> why is this resolving with depth 0? nothing should resolve
> infinitely. yeah this is what i want to avoid

Because the calling code doesn't care about how many steps it takes,
it just wants a resolved path.  If we want to limit the resolution
depth as a policy issue, we just need a way to inject the policy limit
into namesys/base.go's resolve.  That's independent of the calling
code.  In my not-pushed-yet reroll, I've gone with the hard-coded 32
limit mentioned here [1].  Commit looks like:

   // resolve is a helper for implementing Resolver.Resolve using resolveOnce.
   func resolve(ctx context.Context, r resolver, name string, depth int, prefixes ...string) (path.Path, error) {
  +       id depth == 0 {
  +               depth = 32
  +       }
          for {

And I'm punting a serious place to configure that default depth to a
follow-up PR (as mentioned in [1]).

[1]: https://github.com/ipfs/go-ipfs/pull/1208#issuecomment-101526369

jbenet commented on fuse/ipns/ipns_unix.go:
```diff
@@ -150,9 +150,6 @@ func (s *Root) Lookup(ctx context.Context, name string) (fs.Node, error) {
 		return &Link{s.IpfsRoot + "/" + p}, nil
-	} else if segments[0] == "ipns" {
-		p := strings.Join(resolved.Segments()[1:], "/")
-		return &Link{s.IpnsRoot + "/" + p}, nil
```
curious  why this was removed

whyrusleeping replied:
It was likely removed because we do a namesys resolve a few lines up,  and the path will no longer contain ipns after that

wking replied:
On Thu, May 14, 2015 at 09:34:50PM -0700, Juan Batiz-Benet wrote:
> > @@ -150,9 +150,6 @@ func (s *Root) Lookup(ctx context.Context, name string) (fs.Node, error) {
> >  	if segments[0] == "ipfs" {
> >  		p := strings.Join(resolved.Segments()[1:], "/")
> >  		return &Link{s.IpfsRoot + "/" + p}, nil
> > -	} else if segments[0] == "ipns" {
> > -		p := strings.Join(resolved.Segments()[1:], "/")
> > -		return &Link{s.IpnsRoot + "/" + p}, nil
> 
> curious  why this was removed

From 568bdd22b (namesys: Add recursive resolution, 2015-05-07):

  Now that recursive resolution with the namesys/namesys.go composite
  resolver always gets you to an /ipfs/... path, there's no need for
  the /ipns/ special case in fuse/ipns/ipns_unix.go.

Which sounds like what @whyrusleeping was saying.

jbenet commented on namesys/interface.go:
```diff
@@ -31,10 +35,7 @@ type NameSystem interface {
-	Resolve(ctx context.Context, name string) (value path.Path, err error)
-
-	// CanResolve checks whether this Resolver can resolve a name
-	CanResolve(name string) bool
```
why lose this? it's useful for clients to be able to check this.

wking replied:
On Thu, May 14, 2015 at 09:41:42PM -0700, Juan Batiz-Benet wrote:
> > -	// CanResolve checks whether this Resolver can resolve a name
> > -	CanResolve(name string) bool
> 
> why lose this? it's useful for clients to be able to check this.

Not useful enough for us to use it anywhere ;).  Can you think of a
situation where the local check is useful enough that you'd use it?
I'd just try a depth=1 resolve and see if that errored out.

jbenet replied:
Ok



—
Sent from Mailbox

On Fri, May 15, 2015 at 2:13 AM, W. Trevor King <notifications@github.com>
wrote:

>> @@ -31,10 +35,7 @@ type NameSystem interface {
>>  type Resolver interface {
>>  
>>  	// Resolve looks up a name, and returns the value previously published.
>> -	Resolve(ctx context.Context, name string) (value path.Path, err error)
>> -
>> -	// CanResolve checks whether this Resolver can resolve a name
>> -	CanResolve(name string) bool
> On Thu, May 14, 2015 at 09:41:42PM -0700, Juan Batiz-Benet wrote:
>> > -	// CanResolve checks whether this Resolver can resolve a name
>> > -	CanResolve(name string) bool
>> 
>> why lose this? it's useful for clients to be able to check this.
> Not useful enough for us to use it anywhere ;).  Can you think of a
> situation where the local check is useful enough that you'd use it?
> I'd just try a depth=1 resolve and see if that errored out.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ipfs/go-ipfs/pull/1208/files#r30387261

jbenet commented:
@wking  LGTM -- Last few comments o/

wking commented:
Just pushed v3.  Changes since v2:

* “Published to <name>: <value>” [1].
* “Resolve all supported protocols to IPFS” → “Resolve the value
  of names to IPFS” [2].
* “mutable protocols” → “mutable name protocols” [3].
* Default depth of 32 [4].
* Pluggable lookupTXT and DNS-resolution tests [5].

We still need (carried over from the previous summary [6]):

* Docs.  Where should these live?  Is the command-line help sufficent?
  Do I need migration notes about the ‘ipfs name resolve …’ scope
  changes?  Output format changes?
* Possibly more consistent error handling [7].

And we also need:

* To shift the depth configuration to the Resolver interface, and
  store it in our resolver structs, instead of passing it in to
  the Resolve(…) call [8].

[1]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30386229
[2]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30384304
[3]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30384308
[4]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30387103
[5]: https://github.com/ipfs/go-ipfs/pull/1208#issuecomment-101720409
[6]: https://github.com/ipfs/go-ipfs/pull/1208#issuecomment-101526369
[7]: https://github.com/ipfs/go-ipfs/pull/1208#discussion_r29918312
[8]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30384456

Luzifer commented on core/corehttp/ipns_hostname.go:
```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```
In order to prevent a [redirect loop](https://github.com/ipfs/go-ipfs/issues/1233#issuecomment-102618200) here please add Trimming of slashes to this line:

```diff
diff --git a/core/corehttp/ipns_hostname.go b/core/corehttp/ipns_hostname.go
index 7361001..19daefd 100644
--- a/core/corehttp/ipns_hostname.go
+++ b/core/corehttp/ipns_hostname.go
@@ -20,7 +20,7 @@ func IPNSHostnameOption() ServeOption {

                        host := strings.SplitN(r.Host, ":", 2)[0]
                        if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-                               r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+                               r.URL.Path = strings.TrimRight(p.String(), "/") + r.URL.Path
                        }
                        childMux.ServeHTTP(w, r)
                })
```

Luzifer replied:
Other solution which should also fix #1234 would be to change this to `/ipns/` path and let the IPNS stuff do the magic. (For IPNS there is already an exception for `Cache-Control`)

```diff
diff --git a/core/corehttp/ipns_hostname.go b/core/corehttp/ipns_hostname.go
index 7361001..0da667c 100644
--- a/core/corehttp/ipns_hostname.go
+++ b/core/corehttp/ipns_hostname.go
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
                        defer cancel()

                        host := strings.SplitN(r.Host, ":", 2)[0]
-                       if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-                               r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+                       if _, err := n.Namesys.Resolve(ctx, host); err == nil {
+                               r.URL.Path = "/ipns/" + host + r.URL.Path
                        }
                        childMux.ServeHTTP(w, r)
                })
```

Luzifer replied:
I'd prefer the second approach because we don't duplicate the logic here. We just rewrite to a IPNS path which is handled by other logic which also pulls the DNS record and resolves in a better way than we're doing with the first approach. Also for `/ipns/` there are special handlings defined in delivery which would need to be extended for this HostnameOption thing.

wking replied:
On Sat, May 16, 2015 at 10:50:58AM -0700, Knut Ahlers wrote:
> Other solution which should also fix #1234 would be to change this
> to `/ipns/` path and let the IPNS stuff do the magic. (For IPNS
> there is already an exception for `Cache-Control`)

In #1233, the problem is:

  $ curl -IL swagger.luzifer.io

says “ah, I have a hostname, let me try to resolve it via a dnslink”,
and gets:

  /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

Then it attaches the URL path (empty or a slash, doesn't really
matter) and returns a 301 redirect to (modulo the old /ipfs/
injection):

  http://swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

When it serves *that* request, it says “ah, I have a hostname, let me
try to resolve it via a dnslink”, and then attaches the URL path (now
/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA) to redirect to:

  http://swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

which is:

  http://swagger.luzifer.io/{dnslink-from-second-call}/{dnslink-from-first-call}{empty-original-path}

What we're missing is a way to break out of this hostname-prefixing
loop.  I see a few choices here.

a. Defining any URL path to be an absolute path, so:

     $ curl -IL swagger.luzifer.io

   has an empty (enough) path, and would get resolved to
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA, but

     $ curl -IL swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

   has a non-empty path, so it would *also* resolve to
   ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA (and not need a
   redirect).

   But that makes life hard for folks who want pretty names without a
   base-58 multihash [1].

b. Defining any URL path to be relative, so:

     $ curl -IL swagger.luzifer.io

   would just serve up
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA without any
   redirects.  And

     $ curl -IL swagger.luzifer.io/index.html

   would serve up the content referenced by the ‘index.html’ link in
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA (again without
   redirects).

   But that makes life hard for folks to folks who want to use the
   gateway behind swagger.luzifer.io to access content outside of the
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA tree.

c. Come up with some way to signal whether or not your URL path is
   absolute.  So:

     $ curl -IL -H 'IPFS: absolute' swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

   would give you the content of your absolute path, and:

     $ curl -IL -H 'IPFS: relative' swagger.luzifer.io/index.html

   would give you the content of
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html.

   But an IPFS header is pretty ugly ;).

I think the right choice depends on who's running the gateway.  Folks
running an IPFS gateway as a service to the community (or their
subnet) will likely want either (a) or (c), so they can hit arbitrary
IPFS resource.  Folks serving their own website out of an IPFS gateway
will likely want either (b) or (c) so they get pretty URLs (and maybe
so they don't shoulder the bandwidth of an open-to-all-IPFS-space
gateway).  Folks using browsers or writing clients will likely not
want (c), because setting headers can be awkward.  If we have a
smarter way to communicate the absolute/relative distinction, then I
think (c) is the best choice, but I'm not sure what that way would be.
Maybe another leading slash for absolute paths?

  swagger.luzifer.io//ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA
  swagger.luzifer.io/index.html

[1]: https://github.com/ipfs/go-ipfs/issues/1234#issuecomment-102674315

Luzifer replied:
In #1233 the last left problem is not what your thesis states.

The last left problem is we are taking the DNS entry `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/` (dont miss the trailing slash!) and put the request path (`/`) on it. So we have `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA//`which is redirected by the mux (not by any IPFS code) to `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/`. After that we take the DNS entry and attach the path so we have `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA//ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/` which contains a `//` and again the mux is intercepting.

- So one solution would just be ensure there are no `//` in the path after attaching the path to it. (Done that with `strings.TrimRight`)
- The other solution is to use IPNS. (Explained above)

Regarding domains:
- My public gateway (ipfs.hub.luzifer.io) has no `dnslink` TXT entry so this does not apply to those gateways.
- swagger.luzifer.io is a website. Not a gateway. If someone tries to use it as a gateway it should fail by design. Accessing content outside `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA` would be gateway functionality…

wking replied:
On Sat, May 16, 2015 at 02:57:34PM -0700, Knut Ahlers wrote:
> … which is redirected by the mux (not by any IPFS code)…

Ah, I see.  So having the trailing double-slash is a problem for
anything that's getting DNS-link resolution.

> Regarding domains:
> - My public gateway (ipfs.hub.luzifer.io) has no `dnslink` TXT entry
>   so this does not apply to those gateways.
> - swagger.luzifer.io is a website. Not a gateway. If someone tries
>   to use it as a gateway it should fail by design. Accessing content
>   outside `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA`
>   would be gateway functionality…

So that sounds like “the way you distinguish between absolute and
relative paths is whether or not the host has a DNS-link TXT entry”.
I'm fine with that.  And then “do whatever we need to do to avoid
upsetting the muxer,” and the TrimRight makes sense there.

I'd rather not take the /ipns/ approach [1], because then we're
hitting the DNS entry twice (one for “is there a DNS-link” and again
for “resolve this DNS-link-based /ipns/… path”).  And because we've
rolled back my proposed distinction between DNS and DHT links [2],
there's some extra multihash checking that we wouldn't need.  Neither
of those are big deals, but I don't like pulling in logic we know we
won't need here.

[1]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30463056
[2]: https://github.com/ipfs/go-ipfs/pull/1208#issuecomment-100754017

Luzifer replied:
Except the IPNS stuff: ACK

(Let's discuss the IPNS stuff in a second pull-request I'll open after this one has been merged)

jbenet replied:
I was going to say that for `(c)` we do not need to add a header. We already set 

```
X-IPFS-Path: /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html
```

So what the gateway can do is -- after redirecting the first time, setting `X-IPFS-Path` -- it could check for `X-IPFS-Path` and ignore the url's path if so. i.e. stop after (2) below.

Gateway resolve:
1. set `path = /`
2. check X-IPFS-Path, if there is one, set `path = <x-ipfs-path>`. DONE!
3. see hostname -- resolve dnslink. set `path = <dnslink-path>`
4. see URL (relative) path. set `path = path + url.path` DONE!

(not sure if the above is a fully correct characterization in light of @Luzifer's IPNS solution, which we should do).

This -- the `X-IPFS-Path` approach -- is more flexible, i think, as some services may be constrained in their "URL prettiness rules" and the header would be a ground truth of what to resolve.

jbenet commented on namesys/dns_test.go:
```diff
@@ -40,3 +55,74 @@ func TestDnsEntryParsing(t *testing.T) {
+	}
+}
+
+func TestDnsResolution(t *testing.T) {
```
love this test.

whyrusleeping replied:
:+1:

jbenet commented:
@wking quick notes:

- i'd like to be able to contribute to this branch (fix resolve call) and i can't. i'd have to replace the branch with a new PR, which is annoying. (It also _may be_ that Travis isn't running both tests (`pr` and `push`) because it's not a branch in the repo.)  perhaps a good workflow might be: keep your fork, but push the branch to this fork and make the PR from within this fork.

> * Docs.  Where should these live?  Is the command-line help sufficent?
>  Do I need migration notes about the ‘ipfs name resolve …’ scope
>  changes?

Help text should be sufficient for now. We haven't begun to clearly document API changes. this would be one well-documented example among many not. We should move to doing that as we make the shell api and commit to it.

> * To shift the depth configuration to the Resolver interface, and
>   store it in our resolver structs, instead of passing it in to
>   the Resolve(…) call [8].

Let's just do

```go
func (r *resolver) Resolve(ctx context.Context, name string) (path.Path, error) { ... }
func (r *resolver) ResolveWithDepth(ctx context.Context, name string, depth int) (path.Path, error) { ... }
```
and be done.

Have time to make this one quick change today?  I'd do it but i can't push to this branch. I'd like to merge this asap-- if you don't have time right now, i'll just pull this into a new branch and close this PR in favor of another.

jbenet commented:
And, @wking: 

- Great! this LGTM and ready for merge. thanks for all the work cleaning this up.

wking commented:
I've rebased onto master, dropped the 'depth' argument, and expanded
the package godocs.  The only Travis failure looks like an unrelated
bitswap panic, so I think this is mergable now.

> - i'd like to be able to contribute to this branch (fix resolve
>   call) and i can't. i'd have to replace the branch with a new PR,
>   which is annoying.

I don't mind having the branch replaced.  Or you can PR the branch in
my repository.  Or you can just bump me and tell me to get this thing
out the door ;).

> perhaps a good workflow might be: keep your fork, but push the
> branch to this fork and make the PR from within this fork.

Not much point to me having a fork if all the PR branches are coming
from this repository.  If the methods listed above don't sound like
fun, I'll just drop my fork (once any current PRs are merged/closed)
and push to branches in the central repository.

wking commented:
On Sun, May 17, 2015 at 11:54:56AM -0700, W. Trevor King wrote:
> I've rebased onto master, dropped the 'depth' argument, and expanded
> the package godocs.  The only Travis failure looks like an unrelated
> bitswap panic, so I think this is mergable now.

To be on the safe side, I'm working up some tests for the composite
namesys/namesys.go resolver, since we currently only have Go tests for
the single-protocol DHT- and DNS-link resolvers.

wking commented:
Pushed v4.  Changes since v3:

* context.TODO → context.Background [1]
* Added save-hook to run ‘go fmt’ and rebased to incorperate format fixes [2]
* New commit to replace max-int calculation (documented in the
  ‘namesys/interface: Rework Resolve to not take a depth
  argument’ commit message) with additional call to DNS
  resolver's Depth() [3].
* Restored the accidental sadhack removal [4].

[1]: https://github.com/wking/go-ipfs/commit/782b99d4ab6fd47f80879b1e4ed68ade3f7bf06f#commitcomment-11231681
[2]: https://github.com/wking/go-ipfs/commit/782b99d4ab6fd47f80879b1e4ed68ade3f7bf06f#commitcomment-11231685
[3]: https://github.com/wking/go-ipfs/commit/782b99d4ab6fd47f80879b1e4ed68ade3f7bf06f#commitcomment-11231705
[4]: https://github.com/wking/go-ipfs/commit/782b99d4ab6fd47f80879b1e4ed68ade3f7bf06f#commitcomment-11231727

wking commented:
Pushed v5.  Changes since v4:

* Rework IPFS / IPNS distinction to include both mutable and
  immutable objects under “IPFS” [1].
* Reroll the API around Resolve and ResolveN ([2] and offline
  discussion).

[1]: https://github.com/wking/go-ipfs/commit/0067357503b435860e15e51284c5247b88add5aa#commitcomment-11231752
[2]: https://github.com/wking/go-ipfs/commit/782b99d4ab6fd47f80879b1e4ed68ade3f7bf06f#commitcomment-11231745

wking commented:
Pushed v6.  Changes since v5:

* Added namesys/namesys_test.go to excercise mpns.ResolveN.

jbenet commented:
\o/ testing all commits :traffic_light: \o/

![](https://www.evernote.com/l/AMYRsNNiqDpAL7C_SiUZz41tciRRBaSqGhgB/image.png)

jbenet commented:
Actually, we probably should've rebased on new master (fixing travis things) before this...

@wking i'd love to be able to do this on your PRs-- could we move to a workflow where you push a branch to this repo and PR from here?

wking commented:
On Tue, May 19, 2015 at 05:15:39PM -0700, Juan Batiz-Benet wrote:
> Actually, we probably should've rebased on new master (fixing travis
> things) before this...

Done.

> @wking i'd love to be able to do this on your PRs-- could we move to
> a workflow where you push a branch to this repo and PR from here?

I don't mind.  I'll open new PRs from branches based in this repo.

jbenet commented:
Looks like 1d58881 is broken: https://travis-ci.org/ipfs/go-ipfs/builds/63276010 -- the rest all pass. @wking sorry for even more on this PR. I can try to fix it for you later today, if you don't get to it first.

(yeah, this `git push-each` thing turned out to be useful :) )

whyrusleeping commented:
"useful"

:stuck_out_tongue:

wking commented:
On Wed, May 20, 2015 at 06:58:17AM -0700, Juan Batiz-Benet wrote:
> Looks like 1d58881 is broken: https://travis-ci.org/ipfs/go-ipfs/builds/63276010

Likely a rebase error:

  core/commands/resolve.go:78: too many arguments in call to n.Namesys.Resolve

I'll squash a fix an repush shortly.

wking commented:
On Wed, May 20, 2015 at 08:36:19AM -0700, W. Trevor King wrote:
> Wed, May 20, 2015 at 06:58:17AM -0700, Juan Batiz-Benet:
> > Looks like 1d58881 is broken:
> > https://travis-ci.org/ipfs/go-ipfs/builds/63276010
> 
> Likely a rebase error:
> 
>   core/commands/resolve.go:78: too many arguments in call to n.Namesys.Resolve
> 
> I'll squash a fix an repush shortly.

Pushed.

jbenet commented:
### thanks for bearing with me @wking ! :) :clap:

![](http://31.media.tumblr.com/tumblr_m1powhRVdz1rqnp16o1_r1_500.gif)
![](http://2.bp.blogspot.com/-t9SelogaA3E/UM4m_XmWlBI/AAAAAAAAAL4/0-C_AezsP1E/s1600/dancin+in+the+shire+gif.gif)
